core/vnl/vnl_vector.hxx

```diff
@@ -106,96 +119,68 @@ vnl_vector<T>::vnl_vector (vnl_vector<T> const& v)
   }
 }
 
-//: Creates a vector from a block array of data, stored row-wise.
-// Values in datablck are copied. O(n).
-
-template<class T>
-vnl_vector<T>::vnl_vector (T const* datablck, size_t len)
-{
-  vnl_vector_alloc_blah(len);
-  std::copy( datablck, datablck + len, this->data );
-}
-
-//------------------------------------------------------------
-
-template<class T>
-vnl_vector<T>::vnl_vector (vnl_vector<T> const &u, vnl_vector<T> const &v, vnl_tag_add)
-{
-  vnl_vector_alloc_blah(u.num_elmts);
-#if VNL_CONFIG_CHECK_BOUNDS  && (!defined NDEBUG)
-  if (u.size() != v.size())
-    vnl_error_vector_dimension ("vnl_vector<>::vnl_vector(v, v, vnl_vector_add_tag)", u.size(), v.size());
-#endif
-  for (size_t i=0; i<num_elmts; ++i)
-    data[i] = u[i] + v[i];
-}
 
+//: Move-constructs a vector. O(1).
 template<class T>
-vnl_vector<T>::vnl_vector (vnl_vector<T> const &u, vnl_vector<T> const &v, vnl_tag_sub)
+vnl_vector<T>::vnl_vector(vnl_vector<T>&& rhs) noexcept
 {
-  vnl_vector_alloc_blah(u.num_elmts);
-#if VNL_CONFIG_CHECK_BOUNDS  && (!defined NDEBUG)
-  if (u.size() != v.size())
-    vnl_error_vector_dimension ("vnl_vector<>::vnl_vector(v, v, vnl_vector_sub_tag)", u.size(), v.size());
-#endif
-  for (size_t i=0; i<num_elmts; ++i)
-    data[i] = u[i] - v[i];
+  // Copy the data pointer and its length from the source object.
+  this->operator=(std::move(rhs));
 }
 
+//: Move-assigns rhs vector into lhs vector. O(1).
 template<class T>
-vnl_vector<T>::vnl_vector (vnl_vector<T> const &u, T s, vnl_tag_mul)
+vnl_vector<T>& vnl_vector<T>::operator=(vnl_vector<T>&& rhs) noexcept
```

Comment:
Of course, I feel somewhat sad that the move-assignment-operator that I proposed with pull request #693 did not make it. While it was so simple and elegant, just doing a move-construct and a swap.  😢 This one you're proposing here still does dynamic memory allocation, and can still possibly fail, even though it says `noexcept`. I see why you're proposing it this way, but still it looks a bit troublesome.

Reply:
I agree 100%, I view this as a stepping stone move forward.  Removing the reliance on memory allocation in the move constructor is also possible (I think), but we need to stop the willy-nilly implied conversion of types through a vnl_vector_ref that previously always resulted in a memory copy from the assignment operator.  For the moment, this 'hack' in the move operator is to overcome the other design discrepancies.

I think that making all the single argument copy constructors 'explicit' is the next chore for avoiding move-constructor problems.

Here is the crux of the problem that requires memory allocation from from the move assignment.   There are many improvements that need to be done, but my feeling this that this solution is "no worse" than the current situation from a security standpoint, and it provides performance benefits for a large number of instances.

```c++
// psuedocode
{
double mutable_contig_memory[4]{1,2,3,4};
vnl_vector_ref<double> myref( mutable_contig_memory, 4 );
// vnl_vector_ref is-a vnl_vector

vnl_vector<double> M;

M.operator=(  vnl_vector_ref<double>( mutable_contig_memory, 4 ) );
//                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This is an rvalue!
assert(M[3] == 4)
M[3] = 99;
assert(M[3] == 99)
assert(mutable_contig_memory[3] == 99 ); // <- what should the answer be?
assert(mutable_contig_memory[3] == 4 );    // <- what should the answer be?

M.set_size(1000); // <- This will attempt to delete the memory
M.fill(7);
assert(M[77] == 7)
}

{
constexpr double immutable_contig_memory[4]{1,2,3,4};
vnl_vector_ref<double> myref( immutable_contig_memory, 4 );
// vnl_vector_ref is-a vnl_vector

vnl_vector<double> M;

M.operator=(  vnl_vector_ref<double>( immutable_contig_memory, 4 ) );
//                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ This is an rvalue!
assert(M[3] == 4)
M[3] = 99;
assert(M[3] == 99)
// Answer is clear hear: assert(immutable_contig_memory[3] == 99 ); // <- what should the answer be?
assert(immutable_contig_memory[3] == 4 );    // <- what should the answer be?

M.set_size(1000); // <- This will attempt to delete the memory
M.fill(7);
assert(M[77] == 7)
}
```

I really appreciate your feedback.  The current implementation is focused on identifying the scenarios that need to be supported, and then working on intermediate "it-works and passes all tests" solutions that attempt to address only a few issues at a time.

My goal is that this long road will result in your solution after the other deficiencies are addressed.

Reply:
@hjmjohnson I think your code example shows that the current `vnl_vector_ref` logic is very fragile. Right? However, I still think a move-constructor and a move-assignment operator should only be marked `noexcept` if they never fail. Otherwise, I think it's better to remove the `noexcept` keywords from these two member functions.

What do you think of my suggestion to add an extra `SizeSpecifier` template parameter to `vnl_vector<T>`, as an alternative way to allow generic code that supports both fixed and dynamic vnl-vectors? https://github.com/vxl/vxl/pull/694#issuecomment-538502074 Do you think that would be a feasible approach, given the current code base?

Reply:
Yes,  very very fragile.  I don't like it at all.  I like your suggestion a lot, but I don't have the resources or time to take on that project all at once.